share/wake/lib/core/tree.wake

```diff
@@ -13,94 +13,115 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+
 package wake
 
 # balanced order-statistic trees
-data Tree a = Tree (a => a => Order) (TreeNode a) # cmp root
+data Tree a =
+    Tree (a => a => Order) (TreeNode a) # cmp root
 
 from wake export type Tree
 
 data TreeNode a =
-  Tip
-  Bin Integer (TreeNode a) a (TreeNode a) # size left key right
+    Tip
+    Bin Integer (TreeNode a) a (TreeNode a) # size left key right
 
 # delta = 2.5, gamma = 1.5 for a reasonably tight balance
 def deltaQ = 5
+
+
 def deltaD = 2
+
+
 def ratioQ = 3
+
+
 def ratioD = 2
 
+
 # Create a new Tree, sorted by cmp.
 export def tnew cmp = Tree cmp Tip
 
+
 # Convert a List to a Tree.
 export def listToTree cmp list = vectorToTree cmp (listToVector list)
+
+
 export def listToTreeMulti cmp list = vectorToTreeMulti cmp (listToVector list)
 
-export def vectorToTreeMulti cmp v =
-  Tree cmp (build (vsortBy (cmp _ _ | isLT) v))
+
+export def vectorToTreeMulti cmp v = Tree cmp (build (vsortBy (cmp _ _ | isLT) v))
+
 
 # Convert a Vector to a Tree.
 export def vectorToTree cmp v =
-  Tree cmp (build (vdistinctRunBy (cmp _ _ | isEQ) (vsortBy (cmp _ _ | isLT) v)))
+    Tree cmp (build (vdistinctRunBy (cmp _ _ | isEQ) (vsortBy (cmp _ _ | isLT) v)))
+
 
 def build v = match (vlen v)
-  0 = Tip
-  1 = Bin 1 Tip (vat_ 0 v) Tip
-  len =
-    def mid = len >> 1
-    def l = vtake mid v
-    def r = vdrop (mid+1) v
-    Bin len (build l) (vat_ mid v) (build r)
+    0 = Tip
+    1 = Bin 1 Tip (vat_ 0 v) Tip
+    len =
+        def mid = len >> 1
+        def l = vtake mid v
+        def r = vdrop (mid + 1) v
+        Bin len (build l) (vat_ mid v) (build r)
+
 
 # Returns the total length of the Tree.
-export def tlen (Tree _ root: Tree a): Integer =
-  size root
+export def tlen (Tree _ root: Tree a): Integer = size root
+
+
 def size = match _
-  Tip         = 0
-  Bin s _ _ _ = s
+    Tip = 0
+    Bin s _ _ _ = s
+
 
 # Returns True if the Tree is empty, False otherwise.
 export def tempty (Tree _ root: Tree a): Boolean = match root
-  Tip = True
-  _   = False
+    Tip = True
+    _ = False
+
 
 # Insert y into the tree only if no other keys == y
 export def tinsert (y: a) (Tree cmp root: Tree a): Tree a =
-  def helper t = match t
-    Tip         = Bin 1 Tip y Tip
-    Bin _ l x r = match (cmp x y)
-      GT = balanceL (helper l) x r
-      EQ = t
-      LT = balanceR l x (helper r)
-  Tree cmp (helper root)
+    def helper t = match t
+        Tip = Bin 1 Tip y Tip
+        Bin _ l x r = match (cmp x y)
+            GT = balanceL (helper l) x r
+            EQ = t
+            LT = balanceR l x (helper r)
```

Comment:
I'll admit that I feel a little lost in the woods here with complex match cases like this, because there's some jolting of my F#/OCaml and even Elixir senses with `=` replacing `->`, there being no `|` separators, and `match` not being able to be aligned with the match clauses. I know those aren't coming in at all, so not arguing that, but all that to say that I don't think I have much opinion here. For example, despite me knowing Wake syntax, I actually had to read this `Bin _ l x r if cmp x y | isGT =` a few times to re-orient myself that `|` is not a separator. Haha.

@ag-eitilt Could you write out what you were thinking the `match`s could be formatted as? That might help to have something to look at. I think I'm fine with the original match formatting (without thinking about the logic too heavily) given Wake's syntax constraints/choices. If the `match` statements get put on new lines (justifiable), then that unfortunately pushes everything to the right with two additional indentation levels.

Reply:
With us currently leaning toward "line break after every `=`", that comment isn't particularly relevant any more; it was looking at the two `= match` and weighing the relevant benefits of leaving each on the same line.

To go back to my comment somewhere else about other deeply-nested `match` trees, it might wind up cleaner to refactor these with more `require` (this isn't the best example since the second `match` still has three cases and `t` doesn't look great on a line by itself, but I hope you get the idea):
```suggestion
    def helper t =
        require Bin l x r =
            t
        else
            Bin 1 Tip y Tip
        match (cmp x y)
            GT =
                balanceL (helper l) x r
            EQ =
                t
            LT =
                balanceR l x (helper r)
```